### ISSUE TITLE
ADIOS2: Make libffi dependency more relaxed

### ISSUE THREAD

**Issue by ax3l** [LLNL/spack#28930]
### Steps to reproduce the issue

Migrated from https://github.com/libffi/libffi/issues/691#issuecomment-1029310561

@eisenhauer:

> FFS uses FFI, but only in a narrow set of circumstances, including when building on an architecture for which we do not have native code generation. Permutter is x86-64. We don't need and wouldn't use libffi there. I.E. if you just remove FFI from the spack dependency chain, ADIOS, SST and all dependencies should build and work fine. I'm not sure if that kind of conditional dependency can be expressed in spack.

cc @eisenhauer @chuckatkins @pnorbert this libffi build issue breaks the build for ADIOS with SST, which relies on `libffi`, for Perlmutter in the HPE/Cray `PrgEnv-nvidia` programming environment.

### Information on your system

Perlmutter (NERSC) x86-64

Compiling on Linux using the NVHPC (former: PGI) compilers:
(Free download from Nvidia is here: https://developer.nvidia.com/nvidia-hpc-sdk-downloads)

C Compiler: `nvc`
C++ Compiler: `nvc++`

CC E4S: @shahzebsiddiqui

### Additional information

N/A

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by ax3l** [comment]
@eisenhauer We currently have
```py
    depends_on('libffi', when='+sst')            # optional in DILL
```
in the package. We can add `target=aarch64:` et al. to the `when=` clause but we cannot do "not x86" here.

Can you please list me all the supported architecture families where you need and support libffi on? Only `aarch64` and `ppc64le`?

**Comment by eisenhauer** [comment]
Unfortunately, using libffi and emulating DCG is our fallback on all architectures where we _don't_ have direct DCG support.  We have direct DCG support on: 
  -  sparc, ppc64le, x86, x86_64, ia64, arm5, arm6, arm8

and we would need libffi everywhere else.   Generating a list of all unsupported architectures isn't sustainable, so we really need to be able to say "not sparc and not ppc64le and not x86 ..."

**Comment by RemiLacroix-IDRIS** [comment]
Have you found a fix for this? I'm running into the same issue.

**Comment by eisenhauer** [comment]
I don't believe that a fix has been applied, and am not sure a proper fix is possible without extensions to spack (in the form of more complex dependency conditions).  But I could be wrong because I haven't found a lot of documentation on the depends_on() feature...

**Comment by RemiLacroix-IDRIS** [comment]
Is there a way to loop over the possible arch values so that it's easier to add a `depends_on` for any architectures not in sparc, ppc64le, x86, x86_64, ia64, arm5, arm6, arm8?

**Comment by markcmiller86** [comment]
On macOS 12.7.4 anyways, I was able to rid myself of `libffi` dependence using `-DCMAKE_DISABLE_FIND_PACKAGE_LibFFI=TRUE`. It at least built and installed without error. I don't know if the build is somehow invalid as a result.

**Comment by eisenhauer** [comment]
In general, FFS and the software stack above it (EVPath, SST and ADIOS) work fine without libffi.  If it finds libffi additional features are enabled in ffs, but those features are not necessary or useful to the ADIOS software stack.  Newer releases of FFS default to not looking for libffi because the sometimes-dependency has caused trouble.
